src/core/lib/gprpp/time.h

```diff
@@ -70,6 +100,8 @@ class Timestamp {
   static Timestamp FromCycleCounterRoundUp(gpr_cycle_counter c);
   static Timestamp FromCycleCounterRoundDown(gpr_cycle_counter c);
 
+  static Timestamp Now() { return source_->Now(); }
```

Comment:
Every replacement of `ExecCtx::Now` with `Timestamp::Now` in this PR is no longer using cached time. Was that intentional?

Reply:
Every replacement is... ExecCtx installs a new source which is cached, we call the current source (which ExecCtx installed)

Reply:
I see, that's what you meant by "Add a caching time source, and install it in ExecCtx." ... When an ExecCtx is instantiated, the caching time source is installed as the thread's time source via a constructor side effect, and the thread's time source is reset upon ExecCtx destruction.

I have some concern that folks may remove `exec_ctx` instances thinking they aren't being used, and that would silently eliminate time caching. We might not find out until we see some performance decrease and have to backtrack. And call sites wouldn't know whether they need to call `InvalidateNow`, given they don't know if their time is being cached or not. Would it be better to have explicit `Now` and `CachedNow` methods to control the ambiguity?

Reply:
Summary of our chat: Until the promise call end state is reached, 

* folks should not remove ExecCtx instances unless they explicitly mean to eliminate time caching
* callers should assume that their TimeSource is cached, and still call InvalidateNow() when needed (even if it's not actually cached yet).
* eventually, InvalidateNow goes away as a concept.

Reply:
@markdroth can I get a second opinion that you're ok going forward with this? I think it's probably fine, but we'll live in an odd transitional state for a while.

Reply:
I think this is fine.  I suspect that we'll be able to make time caching go away well before `ExecCtx` goes away.

I do agree with AJ that it probably wouldn't be hard to have separate `Now()` and `CachedNow()` methods and identify the few callers that need to use the latter, and if someone wanted to try to do that now without waiting for the promise conversion to finish, I'd be supportive of that.  But either way, this PR seems fine, and whatever order we do things in, we'll get to the right end state.